rclcpp/multithread_executor/multithread_executor.cpp

```diff
@@ -0,0 +1,197 @@
+  /**
+   * A small convenience function for converting a thread ID to a string
+   */
+  std::string string_thread_id(std::thread::id id)
+  {
+    auto hashed = std::hash<std::thread::id>()(id);
+    return std::to_string(hashed);
+  }
+
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  size_t count_;
+};
+
+class DualThreadedNode : public rclcpp::Node
+{
+public:
+  DualThreadedNode()
+  : Node("DualThreadedNode")
+  {
+    /* These define the callback groups
+     * They don't really do much on their own, but they have to exist in order to
+     * assign callbacks to them. They're also what the executor looks for when trying to run multiple threads
+     */
+    callback_group_subscriber1_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+    callback_group_subscriber2_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+
+    // Each of these callback groups is basically a thread
+    // Everything assigned to one of them gets bundled into the same thread
+    auto sub1_opt = rclcpp::SubscriptionOptions();
+    sub1_opt.callback_group = callback_group_subscriber1_;
+    auto sub2_opt = rclcpp::SubscriptionOptions();
+    sub2_opt.callback_group = callback_group_subscriber2_;
+
+    subscription1_ = this->create_subscription<std_msgs::msg::String>(
+      "topic",
+      rclcpp::QoS(10),
+      // std::bind is sort of C++'s way of passing a function
+      // If you're used to function-passing, skip these comments
+      std::bind(
+        &DualThreadedNode::subscriber1_cb,  // First parameter is a reference to the function
+        this,                               // What the function should be bound to
+        std::placeholders::_1),             // At this point we're not positive of all the
+                                            // parameters being passed
+                                            // So we just put a generic placeholder
+                                            // into the binder
+                                            // (since we know we need ONE parameter)
+      sub1_opt);                  // This is where we set the callback group.
+                                  // This subscription will run with callback group subscriber1
+
+    subscription2_ = this->create_subscription<std_msgs::msg::String>(
+      "topic",
+      rclcpp::QoS(10),
+      std::bind(
+        &DualThreadedNode::subscriber2_cb,
+        this,
+        std::placeholders::_1),
+      sub2_opt);
+  }
+
+private:
+  /**
+   * Simple function for generating a timestamp
+   * Used for somewhat ineffectually demonstrating that the multithreading doesn't cripple performace
+   */
+  std::string timing_string()
+  {
+    time_t timer;
+    struct tm * tm_info;
+    struct tm tm_result;
+    char buffer[12];
+
+    time(&timer);
+    tm_info = localtime_r(&timer, &tm_result);
```

Comment:
So this is probably not portable to Windows.  But instead of this dance, this whole method can be something more like:

```
rclcpp::Time time = this->now();
return std::to_string(time.nanoseconds());
```

Reply:
Good catch, thank you.